<a href="https://colab.research.google.com/github/MaxTechniche/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/Jacob_Maxfield_LS_DS20_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [2]:
import pandas as pd
import numpy as np

In [3]:
# Loading in Automobile Dataset
auto_df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data', header=None, na_values='?')
auto_df.head()

0      1            2    3    4     5   ...    20     21      22  23  24       25
0   3    NaN  alfa-romero  gas  std   two  ...   9.0  111.0  5000.0  21  27  13495.0
1   3    NaN  alfa-romero  gas  std   two  ...   9.0  111.0  5000.0  21  27  16500.0
2   1    NaN  alfa-romero  gas  std   two  ...   9.0  154.0  5000.0  19  26  16500.0
3   2  164.0         audi  gas  std  four  ...  10.0  102.0  5500.0  24  30  13950.0
4   2  164.0         audi  gas  std  four  ...   8.0  115.0  5500.0  18  22  17450.0

[5 rows x 26 columns]

In [4]:
column_names = ['symboling', 'normalized-losses', 'make', 'fuel-type', 
                'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 
                'engine-location', 'wheel-base', 'length', 'width', 'height', 
                'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 
                'fuel-system', 'bore', 'stroke', 'compression-ratio', 
                'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']

In [5]:
auto_df.columns = column_names
auto_df.head()

symboling  normalized-losses         make  ... city-mpg highway-mpg    price
0          3                NaN  alfa-romero  ...       21          27  13495.0
1          3                NaN  alfa-romero  ...       21          27  16500.0
2          1                NaN  alfa-romero  ...       19          26  16500.0
3          2              164.0         audi  ...       24          30  13950.0
4          2              164.0         audi  ...       18          22  17450.0

[5 rows x 26 columns]

## 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

### 2) part 1

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and use a 1-sample t-test to compare your estaimated sample mean to the population mean. Can you say that they are different? 

In [6]:
from scipy import stats

In [7]:
# Getting the mean of 'highway-mpg'
h_mpg_mean = auto_df['highway-mpg'].mean()
h_mpg_mean

30.75121951219512

**Actual highway-mpg ~ 30.75**

In [8]:
# Generating a random sample
auto_sample = auto_df.sample(10, random_state=30)
auto_sample

symboling  normalized-losses  ... highway-mpg    price
121          1              154.0  ...          38   6692.0
176         -1               65.0  ...          32  10898.0
24           1              148.0  ...          38   6229.0
88          -1              137.0  ...          30   9279.0
72           3              142.0  ...          18  35056.0
103          0              108.0  ...          25  13499.0
26           1              148.0  ...          38   7609.0
20           0               81.0  ...          43   6575.0
155          0               91.0  ...          32   8778.0
185          2               94.0  ...          34   8195.0

[10 rows x 26 columns]

### Null Hypothesis:

The average highway_mpg of the sample will be that of the average of the lots average.

### Alternative Hypothesis:

The sample highway_mpg mean is not the same as the lots average.

### Confidence Level:

**95% Confidence Level**

In [9]:
# Performing a 1 sample ttest on the data
ttest_value = stats.ttest_1samp(auto_sample['highway-mpg'], h_mpg_mean)
ttest_value

Ttest_1sampResult(statistic=0.8912422391619923, pvalue=0.39600040771093326)

### Conclusion:

Due to a t-statistic of .891 and a p-value of .396, I **fail to reject** the null hypothesis that the sample is a fair representation of the population highway-mpg mean.

### 2) part 2

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

In [ ]:
# I already have my random sample
auto_sample

symboling  normalized-losses  ... highway-mpg    price
121          1              154.0  ...          38   6692.0
176         -1               65.0  ...          32  10898.0
24           1              148.0  ...          38   6229.0
88          -1              137.0  ...          30   9279.0
72           3              142.0  ...          18  35056.0
103          0              108.0  ...          25  13499.0
26           1              148.0  ...          38   7609.0
20           0               81.0  ...          43   6575.0
155          0               91.0  ...          32   8778.0
185          2               94.0  ...          34   8195.0

[10 rows x 26 columns]

### Null Hypothesis:

The salesman's average of about 35 is the lots average.

### Alternative Hypothesis:

The salesman's average of about 35 is not the lots average.

### Confidence Level:

**90% Confidence Level**

In [10]:
# Performing a 1 sample ttest on the sample vs the salesmans average
ttest_value = stats.ttest_1samp(auto_sample['highway-mpg'], 35)
ttest_value

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

### Conclusion

Due to a t-statistic of -.957 and a p-value of .364, I **fail to reject** the null hypothesis that the salesman's average of about 35 is the lots average.

### 2) Part 3

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, take a sample of 100 cars. Run a t-test to verify the salesman's claim with your new larger sample. Do you reach the same conclusion as you did with the sample size of 10? 

In [13]:
# Pulling a sample of 100 and making sure that I do in fact have 100 rows
large_sample = auto_df.sample(100, random_state=30)
large_sample.shape

(100, 26)

In [14]:
# Performing a 1 sample ttest on the larger sample vs the salesmans average
large_ttest_value = stats.ttest_1samp(large_sample['highway-mpg'], 35)
large_ttest_value

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

### Conclusion

Due to a t-statistic of -6.519 and a p-value of < .00001, I **reject** the null hypothesis that the salesman's given average (35) is the lots average.

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

```
The greater the sample, the closer we are to testing the entire population, 
which generally gets us closer to it's average.
```



---


## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [16]:
column_names = ['Class Name', 'handicapped-infants', 'water-project-cost-sharing', 
                'adoption-of-the-budget-resolution', 'physician-fee-freeze',
                'el-salvador-aid', 'religious-groups-in-schools', 'anti-satellite-test-ban',
                'aid-to-nicaraguan-contras', 'mx-missile', 'immigration', 
                'synfuels-corporation-cutback', 'education-spending', 'superfund-right-to-sue',
                'crime', 'duty-free-exports', 'export-administration-act-south-africa']

In [19]:
# Loading in dataset using column_names
congress = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data', names=column_names)
congress

Class Name  ... export-administration-act-south-africa
0    republican  ...                                      y
1    republican  ...                                      ?
2      democrat  ...                                      n
3      democrat  ...                                      y
4      democrat  ...                                      y
..          ...  ...                                    ...
430  republican  ...                                      y
431    democrat  ...                                      y
432  republican  ...                                      y
433  republican  ...                                      y
434  republican  ...                                      n

[435 rows x 17 columns]

In [74]:
# Replacing all y's with 1, n's with 0, and ?'s with NaN
congress = congress.replace({'y': 1, 'n':0, '?':np.NaN, 'nan':np.NaN})
congress

Class Name  ...  export-administration-act-south-africa
0    republican  ...                                     1.0
1    republican  ...                                     NaN
2      democrat  ...                                     0.0
3      democrat  ...                                     1.0
4      democrat  ...                                     1.0
..          ...  ...                                     ...
430  republican  ...                                     1.0
431    democrat  ...                                     1.0
432  republican  ...                                     1.0
433  republican  ...                                     1.0
434  republican  ...                                     0.0

[435 rows x 17 columns]

In [75]:
# Splitting the DataFrame into 2. (Those that are Republican and those that are Democrat)
rep = congress[congress['Class Name'] == 'republican']
dem = congress[congress['Class Name'] == 'democrat']

In [76]:
rep.head()

Class Name  ...  export-administration-act-south-africa
0   republican  ...                                     1.0
1   republican  ...                                     NaN
7   republican  ...                                     1.0
8   republican  ...                                     1.0
10  republican  ...                                     0.0

[5 rows x 17 columns]

In [77]:
dem.head()

Class Name  ...  export-administration-act-south-africa
2   democrat  ...                                     0.0
3   democrat  ...                                     1.0
4   democrat  ...                                     1.0
5   democrat  ...                                     1.0
6   democrat  ...                                     1.0

[5 rows x 17 columns]

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

In [78]:
# I'm just going to do the for loop right now
for issue in congress.columns[1:]:
  test = stats.ttest_ind(dem[issue], rep[issue], nan_policy='omit')
  print(issue)
  print('Null Hypothesis: Democrats and Republicans have the same level of support on this bill.')
  print('Alternative Hypothesis: Democrats and Repubicans do not have the same level of support on this bill.')
  print(test)
  if test.pvalue < .01:
    reject = 'REJECT'
  else:
    reject = 'FAIL TO REJECT'
  print(f'Conclusion: Based on a t-statistic of about {test.statistic:.3f} and a p-value of about {test.pvalue:.3f},\n\t\
    we {reject} the null hypothesis that Democrats and Republicans have the same level of support on this bill.')
  if reject == 'REJECT':
    if test.statistic > 0:
      party = "Democrat's"
    else:
      party = "Republican's"
    print(f'\t{party} support it more.')
  print(f'\tDemocrat issue mean: {round(dem[issue].mean(), 2)}, Republican issue mean: {round(rep[issue].mean(), 2)}')
  print()


handicapped-infants
Null Hypothesis: Democrats and Republicans have the same level of support on this bill.
Alternative Hypothesis: Democrats and Repubicans do not have the same level of support on this bill.
Ttest_indResult(statistic=9.205264294809222, pvalue=1.613440327937243e-18)
Conclusion: Based on a t-statistic of about 9.205 and a p-value of about 0.000,
	    we REJECT the null hypothesis that Democrats and Republicans have the same level of support on this bill.
	Democrat's support it more.
	Democrat issue mean: 0.6, Republican issue mean: 0.19

water-project-cost-sharing
Null Hypothesis: Democrats and Republicans have the same level of support on this bill.
Alternative Hypothesis: Democrats and Repubicans do not have the same level of support on this bill.
Ttest_indResult(statistic=-0.08896538137868286, pvalue=0.9291556823993485)
Conclusion: Based on a t-statistic of about -0.089 and a p-value of about 0.929,
	    we FAIL TO REJECT the null hypothesis that Democrats and Republ

In [79]:
import math

---
## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [80]:
### COMPLETED ABOVE ###



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [100]:
# I'm just going to do the for loop right now
for issue in congress.columns[1:]:
  test = stats.ttest_ind(dem[issue], rep[issue], nan_policy='omit')
  print(issue)
  dem_var = sum([(x-dem[issue].mean())**2 for x in dem[issue] if x in [0, 1]])/dem[issue].count()
  rep_var = sum([(x-rep[issue].mean())**2 for x in rep[issue] if x in [0, 1]])/rep[issue].count()
  t_statistic = (dem[issue].mean()-rep[issue].mean())/math.sqrt(dem_var/dem[issue].count()+rep_var/rep[issue].count())
  print('Based off of aproximation (and continuous rounding), the absolute value of the t-statistic needs to be above about 2.6')
  print(f't-statistic of {t_statistic}')  
  print(test)
  if abs(t_statistic) > 2.6:
    reject = 'REJECT'
  else:
    reject = 'FAIL TO REJECT'
  print(f'Conclusion: Based on a t-statistic of about {t_statistic:.3f},\n\t\
    we {reject} the null hypothesis that Democrats and Republicans have the same level of support on this bill.')
  if reject == 'REJECT':
    if t_statistic > 0:
      party = "Democrat's"
    else:
      party = "Republican's"
    print(f'\t{party} support it more.')
  print(f'\tDemocrat issue mean: {round(dem[issue].mean(), 2)}, Republican issue mean: {round(rep[issue].mean(), 2)}')
  print()


handicapped-infants
Based off of aproximation (and continuous rounding), the absolute value of the t-statistic needs to be above about 2.6
t-statistic of 9.686430008842628
Ttest_indResult(statistic=9.205264294809222, pvalue=1.613440327937243e-18)
Conclusion: Based on a t-statistic of about 9.686,
	    we REJECT the null hypothesis that Democrats and Republicans have the same level of support on this bill.
	Democrat's support it more.
	Democrat issue mean: 0.6, Republican issue mean: 0.19

water-project-cost-sharing
Based off of aproximation (and continuous rounding), the absolute value of the t-statistic needs to be above about 2.6
t-statistic of -0.08919789262856036
Ttest_indResult(statistic=-0.08896538137868286, pvalue=0.9291556823993485)
Conclusion: Based on a t-statistic of about -0.089,
	    we FAIL TO REJECT the null hypothesis that Democrats and Republicans have the same level of support on this bill.
	Democrat issue mean: 0.5, Republican issue mean: 0.51

adoption-of-the-budget